CatalogDataCollection.md

```diff
@@ -0,0 +1,198 @@
+`Object.getOwnPropertyNames()` to discover own properties. The properties are
+visited breadth first after visiting the object's prototype. The graph walker
+stores the following:
+
+1. Type information (object, null, undefined, boolean, number, function);
+2. Prototype information;
+3. Property descriptor metadata (writable, enumerable, etc.);
+4. String representations from `o.toString()` (when available).
+
+### What data are not collected
+
+Plenty of data are missed by this data collection scheme. Examples include:
+
+- Data exposed only in other execution contexts, such as workers;
+- APIs that are not exposed in the default global namespace;
+  - E.g., `window.getSelection()` returns a `Selection`, which is not a
+    constructor that is reachable from the default global namespace
+- APIs that are only exposed under special conditions;
+  - E.g., Parts of `WebUSB` are not available if not USB devices are available
+- APIs that are exposed on instances of objects (not their prototypes), and no
+  such objects exist on our test page.
+
+## Lists of APIs
+
+Each browser release is associated with a list of APIs based on an
+`Object Graph => API List` algorithm. The algorithm, in general works like this:
+
+1. **Store APIs of libraries**: Look for objects that are properties of the
+   global object, such as `window.Math`, and store their own properties;
+2. **Store APIs of constructors**: Look for own properties of functions,
+   excluding parts of the `Function` API;
+3. **Store APIs of `Constructor.prototype`**: Look for `SomeFunction.prototype`
+   and store its own properties as APIs of `SomeFunction`;
+4. **Store APIs of `Constructor.prototype`'s prototypes**: If
+    `SomeFunction.prototype.__proto__` is not `SomeOtherFunction.prototype`,
+    then store its own properties as APIs of `SomeFunction` (same with
+    `SomeFunction.prototype.__proto__.__proto__`, etc.);
+4. **Store APIs of instances**: Some browser versions include important APIs,
+   but only expose them on instances of the object. Luckily, such instances tend
+   to lurk somewhere on the test page's object graph (e.g.,
+   `document.body.style` exposes `CSSStyleDeclaration` on some browser
+   versions). Look for objects that not libraries, constructors, or
+   `Constructor.prototype`s. Store their own properties as APIs on `Constructor`
+   the first time that `instance(.__proto__)^n = Constructor.prototype`.
+
+### Detailed algorithm
+
+In practice, the algorithm relies on a number of heuristics for identifying
+libraries, constructors, prototypes, and interface names. The following captures
+the overall algorithm, with reference to several heuristics that are described
+in more detail below.
+
+```
+ExtractApiCatalog(ObjectGraph):
+
+Let fns    := Map: ObjectGraph id -> [Strings]:
+                   All objects with a "prototype" property mapped to all names
+                   for this object based on "Function naming heuristics"
```

Comment:
Can these be links?

Reply:
Making these links is [painful/difficult inside markdown code blocks](https://stackoverflow.com/questions/20092433/placing-links-inside-markdown-code-blocks).